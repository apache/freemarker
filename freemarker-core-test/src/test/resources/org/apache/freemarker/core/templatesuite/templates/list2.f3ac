<#--
  Licensed to the Apache Software Foundation (ASF) under one
  or more contributor license agreements.  See the NOTICE file
  distributed with this work for additional information
  regarding copyright ownership.  The ASF licenses this file
  to you under the Apache License, Version 2.0 (the
  "License"); you may not use this file except in compliance
  with the License.  You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

  Unless required by applicable law or agreed to in writing,
  software distributed under the License is distributed on an
  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
  KIND, either express or implied.  See the License for the
  specific language governing permissions and limitations
  under the License.
-->
<@testList ["aardvark", "bear", "cat", "dog"] />

<@testList ["aardvark"] />

<@testList [] />

<@testList listables.list />

<@testList listables.set />

<@testList listables.getIterator />

<@testList listables.emptyList />

<@testList listables.emptySet />

<@testList listables.getEmptyIterator />

<#macro testList xs{positional}>
=== [${resolve(xs)?join(", ")}] ===
<#assign resolveCallCnt = 0>

-- List+sep:
<#list resolve(xs) as x>
    ${x}<#sep>,</#sep>
</#list>
-- List+else:
<#list resolve(xs) as x>
    ${x}
<#else>
    Empty!
</#list>
-- List+items:
<#list resolve(xs)>
    [
    <#items as x>
        ${x!'U'}
    </#items>
    ]
</#list>
-- List+items+else:
<#list resolve(xs)>
    [
    <#items as x>
        ${x!'U'}
    </#items>
    ]
<#else>
    Empty!
</#list>
-- List+items+sep+else:
<#list resolve(xs)>
    [
    <#items as x>
        ${x!'U'}<#sep>,</#sep>
    </#items>
    ]
<#else>
    Empty!
</#list>
<@assertEquals expected=5 actual=resolveCallCnt />
--
</#macro>

<#function resolve(xs)>
    <#assign resolveCallCnt = (resolveCallCnt!0) + 1>
    <#if xs?isFunction>
        <#return xs()>
    <#else>
        <#return xs>
    </#if>
</#function>