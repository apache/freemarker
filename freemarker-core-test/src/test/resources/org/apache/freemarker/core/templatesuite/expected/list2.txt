/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
=== [aardvark, bear, cat, dog] ===

-- List+sep:
    aardvark,
    bear,
    cat,
    dog
-- List+else:
    aardvark
    bear
    cat
    dog
-- List+items:
    [
        aardvark
        bear
        cat
        dog
    ]
-- List+items+else:
    [
        aardvark
        bear
        cat
        dog
    ]
-- List+items+sep+else:
    [
        aardvark,
        bear,
        cat,
        dog
    ]
--

=== [aardvark] ===

-- List+sep:
    aardvark
-- List+else:
    aardvark
-- List+items:
    [
        aardvark
    ]
-- List+items+else:
    [
        aardvark
    ]
-- List+items+sep+else:
    [
        aardvark
    ]
--

=== [] ===

-- List+sep:
-- List+else:
    Empty!
-- List+items:
-- List+items+else:
    Empty!
-- List+items+sep+else:
    Empty!
--

=== [11, 22, 33] ===

-- List+sep:
    11,
    22,
    33
-- List+else:
    11
    22
    33
-- List+items:
    [
        11
        22
        33
    ]
-- List+items+else:
    [
        11
        22
        33
    ]
-- List+items+sep+else:
    [
        11,
        22,
        33
    ]
--

=== [11, 22, 33] ===

-- List+sep:
    11,
    22,
    33
-- List+else:
    11
    22
    33
-- List+items:
    [
        11
        22
        33
    ]
-- List+items+else:
    [
        11
        22
        33
    ]
-- List+items+sep+else:
    [
        11,
        22,
        33
    ]
--

=== [11, 22, 33] ===

-- List+sep:
    11,
    22,
    33
-- List+else:
    11
    22
    33
-- List+items:
    [
        11
        22
        33
    ]
-- List+items+else:
    [
        11
        22
        33
    ]
-- List+items+sep+else:
    [
        11,
        22,
        33
    ]
--

=== [] ===

-- List+sep:
-- List+else:
    Empty!
-- List+items:
-- List+items+else:
    Empty!
-- List+items+sep+else:
    Empty!
--

=== [] ===

-- List+sep:
-- List+else:
    Empty!
-- List+items:
-- List+items+else:
    Empty!
-- List+items+sep+else:
    Empty!
--

=== [] ===

-- List+sep:
-- List+else:
    Empty!
-- List+items:
-- List+items+else:
    Empty!
-- List+items+sep+else:
    Empty!
--

